engine/src/main/java/io/camunda/zeebe/engine/processing/common/ElementActivationBehavior.java

```diff
@@ -0,0 +1,245 @@
+          .orElse(Collections.emptyList());
+
+    } else {
+      return elementInstanceState.getChildren(flowScopeKey).stream()
+          .filter(instance -> instance.getValue().getElementIdBuffer().equals(element.getId()))
+          .toList();
+    }
+  }
+
+  private static boolean isProcess(final ExecutableFlowElement element) {
+    return element.getElementType() == BpmnElementType.PROCESS;
+  }
+
+  private long activateFlowScope(
+      final ProcessInstanceRecord processInstanceRecord,
+      final long flowScopeKey,
+      final ExecutableFlowElement flowScope) {
+    final long elementInstanceKey;
+    final long elementInstanceFlowScopeKey;
+
+    if (isProcess(flowScope)) {
+      elementInstanceKey = processInstanceRecord.getProcessInstanceKey();
+      elementInstanceFlowScopeKey = -1L;
+
+    } else {
+      elementInstanceKey = keyGenerator.nextKey();
+      elementInstanceFlowScopeKey = flowScopeKey;
+    }
+
+    activateFlowScopeByEvents(
+        processInstanceRecord, flowScope, elementInstanceKey, elementInstanceFlowScopeKey);
+
+    return elementInstanceKey;
+  }
+
+  private void activateFlowScopeByEvents(
+      final ProcessInstanceRecord processInstanceRecord,
+      final ExecutableFlowElement element,
+      final long elementInstanceKey,
+      final long flowScopeKey) {
+
+    final var elementRecord = createElementRecord(processInstanceRecord, element, flowScopeKey);
+
+    stateWriter.appendFollowUpEvent(
+        elementInstanceKey, ProcessInstanceIntent.ELEMENT_ACTIVATING, elementRecord);
+    stateWriter.appendFollowUpEvent(
+        elementInstanceKey, ProcessInstanceIntent.ELEMENT_ACTIVATED, elementRecord);
+
+    createEventSubscriptions(element, elementRecord, elementInstanceKey);
+  }
+
+  private long activateElementByCommand(
+      final ProcessInstanceRecord processInstanceRecord,
+      final AbstractFlowElement elementToActivate,
+      final long flowScopeKey) {
+
+    final var elementInstanceKey = keyGenerator.nextKey();
+    final var elementRecord =
+        createElementRecord(processInstanceRecord, elementToActivate, flowScopeKey);
+    commandWriter.appendFollowUpCommand(
+        elementInstanceKey, ProcessInstanceIntent.ACTIVATE_ELEMENT, elementRecord);
+
+    return elementInstanceKey;
+  }
+
+  private ProcessInstanceRecord createElementRecord(
+      final ProcessInstanceRecord processInstanceRecord,
+      final ExecutableFlowElement elementToActivate,
+      final long flowScopeKey) {
+    final var elementInstanceRecord = new ProcessInstanceRecord();
+    // take the properties from the process instance
+    elementInstanceRecord.wrap(processInstanceRecord);
+    // override the properties for the specific element
+    elementInstanceRecord
+        .setElementId(elementToActivate.getId())
+        .setBpmnElementType(elementToActivate.getElementType())
+        .setFlowScopeKey(flowScopeKey)
+        .setParentProcessInstanceKey(-1L)
+        .setParentElementInstanceKey(-1L);
```

Comment:
❓ Nothing to do with this feature at all, but just for my understanding. Am I right that this parent element instance key would refer to the element instance of the call activity task of the parent process?

Reply:
Yes, exactly :dart: